packages/lesswrong/server/resolvers/ultraFeedResolver.ts

```diff
@@ -64,85 +64,54 @@ export const ultraFeedGraphQLTypeDefs = gql`
   }
 `
 
+// items now carry `type`
+type SampledItem =
+  | { type: "feedCommentThread"; feedCommentThread: FeedCommentsThread }
+  | { type: "feedPostWithContents"; feedPost: FeedFullPost }
+  | { type: "feedPost"; feedPostStub: FeedPostStub }          // ← NEW stub
+  | { type: "feedSpotlight"; feedSpotlight: FeedSpotlight };
+
 interface WeightedSource {
   weight: number;
-  items: FeedItem[];
-  renderAsType: FeedItemRenderType;
+  items: SampledItem[];
 }
 
-type SampledItem = { renderAsType: "feedCommentThread", feedCommentThread: FeedCommentsThread }
-                 | { renderAsType: "feedPost", feedPost: FeedFullPost }
-                 | { renderAsType: "feedSpotlight", feedSpotlight: FeedSpotlight };
-
 const weightedSample = (
-  inputs: Record<FeedItemSourceType, WeightedSource>,
+  inputs: Partial<Record<FeedItemSourceType, WeightedSource>>,
   totalItems: number
 ): SampledItem[] => {
-  // Create deep copies of the input arrays to avoid modifying the originals
   const sourcesWithCopiedItems = cloneDeep(inputs);
-
   const finalFeed: SampledItem[] = [];
-  let totalWeight = Object.values(sourcesWithCopiedItems).reduce(
-    (sum, src) => sum + (src.items.length > 0 ? src.weight : 0),
-    0
-  );
+
+  let totalWeight = Object.values(sourcesWithCopiedItems)
+    .reduce((sum, s) => sum + ((s && s.items.length) ? s.weight : 0), 0);
 
   for (let i = 0; i < totalItems; i++) {
     if (totalWeight <= 0) break;
-
     const pick = Math.random() * totalWeight;
-
     let cumulative = 0;
-    let chosenSourceKey: FeedItemSourceType | null = null;
-
-    for (const [key, src] of Object.entries(sourcesWithCopiedItems)) {
-      if (src.items.length === 0) continue;
+    let chosen: FeedItemSourceType | null = null;
 
+    for (const [k, src] of Object.entries(sourcesWithCopiedItems)) {
+      if (!src || !src.items.length) continue;
       cumulative += src.weight;
-      if (pick < cumulative) {
-        chosenSourceKey = key as FeedItemSourceType;
-        break;
-      }
+      if (pick < cumulative) { chosen = k as FeedItemSourceType; break; }
```

Comment:
did anything change here?  (very strongly disprefer multiple lines of code on one line.)